### PR TITLE
Add readyPromise to ImageryProviders

### PR DESCRIPTION
**Pull request by klim705**
Added promise property (similar to `Model` and `Primitive`'s `readyPromise`) to each imagery provider that resolves to true when the provider is ready for use.

### CODE REVIEW

**Comment by hpinkos** [comment]
@klim705 this looks good to me!  Are you going to add a `readyPromise` to the terrain providers as well?

**Comment by klim705** [comment]
@hpinkos It's not necessary at the moment, so I can add it to the terrain providers once we get there, if needed. Thank you!

**Comment by mramato** [comment]
Actually, @hpinkos makes a good point, for completeness we should have readyPromises for terrain as well, I think this will take care of everywhere in Cesium we currently have a `ready` boolean.

**Comment by klim705** [comment]
Okay, I'll just add it to terrain as well and add to this PR then.

**Comment by hpinkos** [comment]
Great! Let me know when it's ready and I'll take a look.

**Comment by klim705** [comment]
Added to terrain providers. I found the `ready` property without a `readyPromise` in `QuadtreeTileProvider`, `GlobeSurfaceTileProvider`, and `Billboard` as well. Should a `readyPromise` also be added to those three?

**Comment by hpinkos** [comment]
Lots of people have asked for a `readyPromise` for `Billboard`.  I think that is out of the scope of this PR though, so I wouldn't worry about it for here.

**Comment by mramato** [comment]
Billboard is a different beast, so I wouldn't worry about it.  The other ones are private, so no need to worry about them either.

File: Source/Core/ArcGisImageServerTerrainProvider.js [hunk 1]
```diff
@@ -152,6 +152,18 @@ define([
...
 
         /**
+         * Gets a promise that resolves to true when the provider is ready for use.
+         * @memberof ArcGisImageServerTerrainProvider.prototype
+         * @type {Promise.<Boolean>}
+         * @readonly
+         */
+        readyPromise : {
+            get : function() {
+                return when.resolve(true);
```
Comment (mramato) [inline Source/Core/ArcGisImageServerTerrainProvider.js]: You should define `this._readyPromise = when.resolve(true);` in the constructor and then simply return it here.  The code as it is now returns a different promise every time.

File: Source/Core/EllipsoidTerrainProvider.js [hunk 2]
```diff
@@ -111,6 +113,18 @@ define([
...
 
         /**
+         * Gets a promise that resolves to true when the provider is ready for use.
+         * @memberof EllipsoidTerrainProvider.prototype
+         * @type {Promise.<Boolean>}
+         * @readonly
+         */
+        readyPromise : {
+            get : function() {
+                return when.resolve(true);
```
Comment (mramato) [inline Source/Core/EllipsoidTerrainProvider.js]: Same as my first comment, I won't mention it anymore, just change it everywhere you were doing this.

File: Source/Core/TerrainProvider.js [hunk 3]
```diff
@@ -1,10 +1,12 @@
 /*global define*/
 define([
+        '../ThirdParty/when',
         './defined',
         './defineProperties',
         './DeveloperError',
         './Math'
     ], function(
+        when,
```
Comment (mramato) [inline Source/Core/TerrainProvider.js]: This class doesn't need to include `when`.

File: Source/Scene/WebMapServiceImageryProvider.js [hunk 4]
```diff
@@ -27,6 +28,7 @@ define([
         queryToObject,
         WebMercatorTilingScheme,
         Uri,
+        when,
```
Comment (mramato) [inline Source/Scene/WebMapServiceImageryProvider.js]: `when` isn't used in this file.

File: Source/Core/CesiumTerrainProvider.js [hunk 5]
```diff
@@ -601,6 +603,18 @@ define([
...
 
         /**
+         * Gets a promise that resolves to true when the provider is ready for use.
+         * @memberof CesiumTerrainProvider.prototype
+         * @type {Promise.<Boolean>}
+         * @readonly
+         */
+        readyPromise : {
+            get : function() {
+                return this._readyPromise;
```
Comment (mramato) [inline Source/Core/CesiumTerrainProvider.js]: Since we are using a deferred, this should be `this._readyPromise.promise` see #3152 for why.

**Comment by mramato** [comment]
One last thing, in providers that aren't always ready, I'm pretty sure there is an error handler somewhere that spits out a console message on failure.  We should `reject` the promise with the same error in these situations (as a `RuntimeError` with the same message).

**Comment by mramato** [comment]
Other than that, this looks great.  Thanks @klim705! and congrats on your first Cesium PR.

**Comment by klim705** [comment]
Finished with review changes. Thanks @hpinkos and @mramato!

**Comment by mramato** [comment]
@klim705 can you add some tests to make sure the promise gets rejected for providers where that can happen?  Other than that this looks great and I'll merge as soon as those changes are in.

**Comment by klim705** [comment]
@mramato Added rejection tests.

**Comment by mramato** [comment]
Thanks again @klim705.

**Comment by mramato** [comment]
I updated CHANGES in master.

**Pull request merged**